### ISSUE TITLE
Unable to zoom in all the way on satellite/aerial imagery

### ISSUE THREAD

**Issue by Brendan-D15** [Helium314/StreetComplete#615]
Zooming in all the way on the map doesn't render tiles at the highest zoom available.

**How to Reproduce**
Zoom in all the way. Compare highest zoom at any location with what can be achieved using the same source in a browser/other tools.

This happens with both the default tile source, and other sources 
ex:
https://clarity.maptiles.arcgis.com/arcgis/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}

**Expected Behavior**
Zooming in all the way should render the tiles at the highest zoom available.

**Does it happen in normal StreetComplete?**
Normal StreetComplete doesn't show satellite imagery (that I could find).

**Versions affected**
Android Version: 14
SCEE version: 58.21

**Comment by Helium314** [comment]
You would need to specifiy which zoom levels are available, which is not possible currently. For simplicity SC default value is used.

**Comment by mic140** [comment]
What does "SC default value" mean? There is no option to connect satellite images there. The problem appeared in recent versions. In earlier versions this problem did not exist.

**Comment by mic140** [comment]
It seems that there is a limit on the maximum zoom. It should either be removed or increased to at least 18

**Comment by mnalis** [comment]
> The problem appeared in recent versions. In earlier versions this problem did not exist.

Can you specify in which version the problem did not exist? (even better, if you can triage which was the first SCEE version where the problem appeared?)

(A screenshot of the same area working at maximum zoom in older SCEE version and not working with newer one might be helpful to clarify the situation)

**Comment by mic140** [comment]
I did some research. The problem definitely wasn't in version 58.21. It appeared in version 59.0 alpha 1
![Screenshot_2024-08-25-06-09-31-035_de westnordost streetcomplete expert](https://github.com/user-attachments/assets/846aaecf-7f56-4498-b17a-1d3a6d2ddf26)
![Screenshot_2024-08-25-06-04-11-863_de westnordost streetcomplete exper2](https://github.com/user-attachments/assets/5989b710-d0b1-4c00-8934-23e4e9b00898)
There is also an unpleasant feature that can be seen on the screenshots - silhouettes of buildings that cannot be turned off. They sometimes interfere with seeing the details of the image.

**Comment by Helium314** [comment]
> What does "SC default value" mean?

Vector tiles used by SC also have a max zoom level

**Comment by Helium314** [comment]
> It should either be removed or increased to at least 18

Have you actually tried to zoom to level 18 with the default raster tile source? E.g. looked at the beautiful imagery in https://server.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/18/141833/93384

**Comment by Helium314** [comment]
> The problem definitely wasn't in version 58.21. It appeared in version 59.0 alpha 1

If there is a difference to the previous version that should be fixable by finding the right place, and the previous maxzoom level.

**Comment by mic140** [comment]
> > The problem definitely wasn't in version 58.21. It appeared in version 59.0 alpha 1
> 
> If there is a difference to the previous version that should be fixable by finding the right place, and the previous maxzoom level.

I didn't understand what you meant. Isn't it obvious from the screenshots that the same place on the second screenshot is clearly not downloaded?

**Comment by mic140** [comment]
By the way, there is no such problem in version 58.22

**Comment by mic140** [comment]
> > It should either be removed or increased to at least 18
> 
> Have you actually tried to zoom to level 18 with the default raster tile source? E.g. looked at the beautiful imagery in https://server.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/18/141833/93384

I wonder where you found such a tile? Mine shows this: https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/18/85660/157174

**Comment by Brendan-D15** [comment]
For a specific example, around Disney World (Orlando, Florida), from the app:

![IMG_20240825_072438](https://github.com/user-attachments/assets/ec15693b-88a0-4d9b-b0e6-5c101eb93198)

This same tile (highlighted in yellow) can be seen with a zoom of 21 (from the source): 
![573327](https://github.com/user-attachments/assets/08287aac-0d96-45fb-807b-344d57dca7e0)

https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/21/875883/573327

**Comment by mic140** [comment]
The release of version 59.0 did not change the situation

**Comment by RubenKelevra** [comment]
I can confirm this bug for Version 59.2, the aerials are basically useless.

**Comment by Helium314** [comment]
This is what I got at zoom 21:
![Screenshot_20241227-115903_SCEE_Dev](https://github.com/user-attachments/assets/5df73c3f-8399-4053-9fcb-6fcee7117060)
This is **actually** useless, not some it's-not-the-best-so-I-call-it-useless kind of useless. And I'm not going to do a change where even in an area with reasonably good aerial imagery user may see something like this.

**Comment by RubenKelevra** [comment]
@Helium314 I can't follow your argument here. You're saying everyone must now live with completely useless aerials at what, zoom level 17 or so because in some areas there's no zoom level 21 available?

Here's the state of Aerials in SCEE. I don't see any use case where looking at this would be useful.


![Screenshot_2024-12-27-12-35-37-595-edit_de westnordost streetcomplete expert](https://github.com/user-attachments/assets/6823199a-9137-4693-9e91-0f9a98da2531)

**Comment by mnalis** [comment]
> If there is a difference to the previous version that should be fixable by finding the right place, and the previous maxzoom level.

This seems to be the location: https://github.com/mnalis/StreetComplete/commit/caf9da60543756ce3f807cd63b252ae0f0c982b6#diff-41f6be370d630835820f7f9484bc80868d12a5b58d6b75d2c976d0dbd85b5db0L662

<details>
<summary>Screenshot  in SCEE 58.2 and master with that patch (click to expand...)</summary>

Regular SCEE 58.2 (zoom in to the maximum):
![small_Screenshot_20241227_185142_SCEE](https://github.com/user-attachments/assets/83d44dfb-cb9b-48e1-a662-4d055479f3fe)

Dev version with patch of `maxzoom=21`:
 ![small_Screenshot_20241227_185148_SCEE Dev](https://github.com/user-attachments/assets/8b0832dd-5150-4c64-898e-209490f10e5a)


</details>

If people want debug APK to try how it works for them, it can be found [here](https://github.com/mnalis/StreetComplete/actions/runs/12517753154).

21 it too much of zoom level however for default imagery, because if I zoom in more, the satellite images disappear.

I can try to find what was the previous value, but it would be _**nicest if one could manually specify the maxzoom**_ in `Tile source for satellite / aerial images`  (as different URLs will have different maxzoom, and we can't autodetect it for TMS AFAIK)


>  There is also an unpleasant feature that can be seen on the screenshots - silhouettes of buildings that cannot be turned off. They sometimes interfere with seeing the details of the image.

@mic140 Are those buildings shown only in SCEE master (and not in latest 58.2 release)?

Anyway, you can go in `Settings` / `Data management` / `Tile source for satellite / aerial images` and turn on the `Hide labels` . That should remove the OSM maps, and show only Satellite imagery so you can have the clearest view. (and if you want some of the things to be displayed but not others, you can enable custom overlay specifying exactly what you want to see. And overlay is easy to turn off/on quickly, esp. if you use `Settings` / `UI` / `Show overlay quick selector`)

**Comment by mic140** [comment]
> That should remove the OSM maps, and show only Satellite imagery so you can have the clearest view

Yes, thank you, it helped, but with the current quality of displaying images, their use is almost useless. I don't understand why it is impossible to simply return the way satellite images were displayed up to and including version 58.22? After all, there is not even a 17 zoom there now. It would be nice with it.

**Comment by mic140** [comment]
> If people want debug APK to try how it works for them, it can be found [here](https://github.com/mnalis/StreetComplete/actions/runs/12517753154).

This is what is needed! If this is included in the main version, it will be great. And the fact that in some places, starting from a certain zoom, the image disappears - this is not as tragic as what is now in the main version. And of course, the ability to save several layers of satellite images and switch between them is very much lacking.

**Comment by mic140** [comment]
> This is what I got at zoom 21: ![Screenshot_20241227-115903_SCEE_Dev](https://private-user-images.githubusercontent.com/43007630/398922343-5df73c3f-8399-4053-9fcb-6fcee7117060.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0OTc2NjUsIm5iZiI6MTczNTQ5NzM2NSwicGF0aCI6Ii80MzAwNzYzMC8zOTg5MjIzNDMtNWRmNzNjM2YtODM5OS00MDUzLTlmY2ItNmZjZWU3MTE3MDYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI5VDE4MzYwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxYWFlODk3ZjlmMjc1ZTdlZGQzZTFmNDUyZTU1OTkwOTc5ZmUzOTA4NDFkNzhkZTk4NDg5NmQ1ZjkwNjZkYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.lSLo1B2q4P7jCGHUz6UYuyFjQaVY3a-vv0buBjNqhDc) This is **actually** useless, not some it's-not-the-best-so-I-call-it-useless kind of useless. And I'm not going to do a change where even in an area with reasonably good aerial imagery user may see something like this.

Why not add the ability to enable this mode in the settings? Let the user choose what is most convenient for him. For example, I am quite satisfied with how it is implemented here: https://github.com/mnalis/StreetComplete/actions/runs/12517753154

**Comment by mnalis** [comment]
> Why not add the ability to enable this mode in the settings? 

@mic140 @Brendan-D15 @RubenKelevra I've tried to implement custom maxzoom (defaulting to 21) on that satellite/aerial imagery URL.

Debug .APK (can be run alongside regular SCEE) is available at  https://github.com/Helium314/SCEE/pull/708 - would you please try it out and let me know how it works for you?

**Comment by mcliquid** [comment]
> would you please try it out and let me know how it works for you?

Works for me!

**Comment by RubenKelevra** [comment]
@mnalis this version works fine:

![Screenshot_2024-12-30-11-17-00-683-edit_de.westnordost.streetcomplete.expert.debug.jpg](https://github.com/user-attachments/assets/30ea8ee2-916a-4537-98cc-3dfeec4dc390)

vs the broken Version 60:

![Screenshot_2024-12-30-11-07-15-440-edit_de.westnordost.streetcomplete.expert.jpg](https://github.com/user-attachments/assets/8173d6f7-9959-46da-9c4f-a48bdb8eb2e6)

**Comment by mic140** [comment]
> would you please try it out and let me know how it works for you?

This option is already quite normal. However, I got the impression that at zooms smaller than the maximum, the application does not load the image corresponding to the current zoom. This is visible in the screenshots from two applications - SCEE and Osmand:
![Screenshot_2024-12-30-13-19-27-954_de westnordost streetcomplete expert debug](https://github.com/user-attachments/assets/ce81a9fd-0a86-4d99-a4ae-70a50a9355f6)
![Screenshot_2024-12-30-13-19-14-868_net osmand plus](https://github.com/user-attachments/assets/cac5a66d-fa46-4cea-a14c-527fe34342d4)

**Comment by Brendan-D15** [comment]
> > Why not add the ability to enable this mode in the settings?
> 
> @mic140 @Brendan-D15 @RubenKelevra I've tried to implement custom maxzoom (defaulting to 21) on that satellite/aerial imagery URL.
> 
> Debug .APK (can be run alongside regular SCEE) is available at #708 - would you please try it out and let me know how it works for you?

Oh yeah, this is a lot better. Thanks!

I tested out setting the maxzoom too high and too low and it works great. Zooming in too much results in the "map data not available yet" warning tiles. Ideally this would be just a label that shows up on top of the last rendered tile, but it's better than what it was.

Zooming back out is ok. It would be nice if it held on to the max zoom tiles a bit longer than how it handles it now... But that's another issue entirely. Also, trying to modify that is not going to go over well until the "map data not available yet" (overzoom issue) is handled a bit differently.

**Comment by mnalis** [comment]
> Zooming in too much results in the "map data not available yet" warning tiles. Ideally this would be just a label that shows up on top of the last rendered tile, but it's better than what it was.

Yeah, unfortunately that is [verbatim image](https://server.arcgisonline.com/arcgis/rest/services/World_Imagery/MapServer/tile/18/141833/93384) that TMS server returns when overzoomed, and not much that SCEE could do about that. I've suggested some heuristics how it might be detected and  better handled by MapLibre in https://github.com/maplibre/maplibre-native/issues/3110, so feel free to provide comments/likes/alternative ideas there.

**Comment by mnalis** [comment]
> However, I got the impression that at zooms smaller than the maximum, the application does not load the image corresponding to the current zoom. This is visible in the screenshots from two applications - SCEE and Osmand:

@mic140 huh, interesting. However, if that is the case, it seems like general MapLibre issue (i.e. I do not think SCEE has control over when MapLibre will start fetching which zoom level), so should be reported in [their issue tracker](https://github.com/maplibre/maplibre-native/issues/). Please do link back to here if you open the issue there, so that interested parties may follow.
